dbms/src/Common/NotNull.h

```diff
@@ -0,0 +1,54 @@
+// Copyright 2024 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <Common/Exception.h>
+
+#include <memory>
+#include <not_null.hpp>
+
+namespace DB
+{
+
+template <typename T>
+using NotNull = cpp::bitwizeshift::not_null<T>;
+
+// It involves an extra check when constructing,
+// if we can't tell if `ptr` is nullptr at compile time.
+template <class T>
+auto newNotNull(T && ptr) noexcept
+{
+    return cpp::bitwizeshift::check_not_null(std::move(ptr));
+}
```

Comment:
Maybe use those functions somewhere to give an example.

Reply:
In `TEST(NotNullTest, Raw)`